File: 0853-car-fleet/0853-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        size_t n = position.size();

        vector<pair<int,int>> cars;

        for(int i=0; i<n; i++) cars.push_back({position[i], speed[i]});

        sort(cars.begin(), cars.end());

        stack<double> time;

        for(int i=n-1; i>=0; i--){
            double t = static_cast<double>(target - cars[i].first) / cars[i].second;

            if(time.empty() || t > time.top()) time.push(t);
        }

        return time.size();
    }
};

File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pairs = [(p,s) for p,s in zip(position,speed)]
        pairs.sort(key=lambda x: x[0], reverse=True)
        max_,fleet = -math.inf, 0
        for p,s in pairs:
            time = (target - p) / s
            if time > max_:
                max_ = time
                fleet += 1

        return fleet